promql/histogram_stats_iterator.go

```diff
@@ -19,32 +19,39 @@ import (
 	"github.com/prometheus/prometheus/tsdb/chunkenc"
 )
 
-type histogramStatsIterator struct {
+type HistogramStatsIterator struct {
 	chunkenc.Iterator
 
 	currentH *histogram.Histogram
 	lastH    *histogram.Histogram
 
 	currentFH *histogram.FloatHistogram
 	lastFH    *histogram.FloatHistogram
+
+	haveReadFromCurrentSeries bool
 }
 
 // NewHistogramStatsIterator creates an iterator which returns histogram objects
 // which have only their sum and count values populated. The iterator handles
 // counter reset detection internally and sets the counter reset hint accordingly
 // in each returned histogram objects.
-func NewHistogramStatsIterator(it chunkenc.Iterator) chunkenc.Iterator {
-	return &histogramStatsIterator{
+func NewHistogramStatsIterator(it chunkenc.Iterator) *HistogramStatsIterator {
+	return &HistogramStatsIterator{
 		Iterator:  it,
 		currentH:  &histogram.Histogram{},
 		currentFH: &histogram.FloatHistogram{},
 	}
 }
 
+func (f *HistogramStatsIterator) Reset(it chunkenc.Iterator) {
```

Comment:
Not changed in this PR, but why is the receiver called `f` rather than `i` or at least something that has something to do with `HistogramStatsIterator`?

Reply:
This deserves a doc comment

And maybe this needs to set `lastH` and `lastFH` to `nil`, at least with #16682 in?

Reply:
> This deserves a doc comment

Added in 55649f8.

> And maybe this needs to set `lastH` and `lastFH` to `nil`, at least with #16682 in?

We could do that, and remove `haveReadFromCurrentSeries`, but then we'd allocate a new `Histogram` or `FloatHistogram` after every `Reset`. If we retain `lastH` and `lastFH`, we can safely reuse them.